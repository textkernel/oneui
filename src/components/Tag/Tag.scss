.Tag {
    display: flex;
    align-items: center;

    padding: var(--spacing-normal) var(--spacing-2x);
    border-radius: var(--border-radius);

    &--isSelected {
        box-shadow: 0 0 0 2px var(--color-primary);
    }

    &--clickable {
        cursor: pointer;
    }

    &__text{
        margin: 0;

        white-space: nowrap;
        overflow: hidden;
        text-overflow: ellipsis;
    }

    &__deleteButton {
        display: flex;
        align-items: center;

        min-width: 15px;
        min-height: 15px;

        border: 0;
        cursor: pointer;
        background-color: var(--transparent);

        padding: 0;
        margin-left: var(--spacing-normal);
    }
}
