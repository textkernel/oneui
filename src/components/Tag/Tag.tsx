import * as React from 'react';
import { MdClose } from 'react-icons/md';
import { bem } from '../../utils';
import { Text } from '../Text';
import { ENTER_KEY } from '../../constants';
import styles from './Tag.scss';

interface Props extends React.HTMLAttributes<HTMLDivElement> {
    /** Value that helps us to understand if the component is currently selected or not. */
    isSelected?: boolean;
    /** Color is assigned to a background color of a wrapper */
    bgColor?: string;
    /** Max-width of a component */
    maxWidth?: string;
    /** Size of the text */
    size?: 'small' | 'normal' | 'large';
    /** Callback, that is fired when a user clicks on a delete icon */
    onDelete?: () => void;
    /** Callback, that is fired when a user clicks on an element */
    onClick?: () => void;
}

const { block, elem } = bem('Tag', styles);

export const Tag: React.FC<Props> = (props) => {
    const { children, bgColor, maxWidth, size, onDelete, onClick, isSelected, ...rest } = props;

    // Generate title for children that are plain text (without tags)
    // If there is something different from the string (JSX) - children will be of object type.
    const areChildrenString = typeof children === 'string';

    const handleDeleteClick = (e: React.MouseEvent<HTMLElement>) => {
        e.stopPropagation();

        onDelete?.();
    };

    const handleTagKeyPress = (e: React.KeyboardEvent) => {
        e.stopPropagation();

        if (e.key === ENTER_KEY) {
            onClick?.();
        }
    };

    const handleDeleteButtonKeyPress = (e: React.KeyboardEvent) => {
        e.stopPropagation();
        e.preventDefault();

        if (e.key === ENTER_KEY) {
            onDelete?.();
        }
    };

    return (
        <div
            {...rest}
            {...block({ ...props, clickable: !!onClick })}
            {...(onClick && {
                onClick,
                tabIndex: 0,
                role: 'button',
                onKeyPress: handleTagKeyPress,
            })}
            style={{
                backgroundColor: bgColor,
                maxWidth,
            }}
        >
            <Text size={size} {...(areChildrenString && { title: children })} {...elem('text')}>
                {children}
            </Text>
            {onDelete && (
                <button
                    onClick={handleDeleteClick}
                    onKeyPress={handleDeleteButtonKeyPress}
                    type="button"
                    {...elem('deleteButton')}
                >
                    <MdClose size="15px" />
                </button>
            )}
        </div>
    );
};

Tag.displayName = 'Tag';

Tag.defaultProps = {
    isSelected: false,
    bgColor: 'var(--color-background)',
    maxWidth: 'fit-content',
    size: 'normal',
    onDelete: undefined,
    onClick: undefined,
};
