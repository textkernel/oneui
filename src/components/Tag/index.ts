export { Tag, Props as TagProps } from './Tag';
