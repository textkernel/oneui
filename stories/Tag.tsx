import * as React from 'react';
import { Tag } from '@textkernel/oneui';

export default {
    title: 'Atoms/Tag',
    component: Tag,
    argTypes: {
        bgColor: {
            options: ['#3eff2b', '#ffa139', 'var(--color-background)'],
        },
        size: {
            options: ['small', 'normal', 'large'],
        },
    },
};

export const _Tag = (args) => {
    return (
        <div
            style={{
                padding: '5px',
            }}
        >
            <Tag {...args} />
        </div>
    );
};
_Tag.args = {
    bgColor: '#3eff2b',
    isSelected: false,
    maxWidth: 'fit-content',
    size: 'normal',
    children: 'This is an extremely long long text!',
};
