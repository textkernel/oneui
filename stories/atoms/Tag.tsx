import * as React from 'react';
import { FiCheck } from 'react-icons/fi';
import { Tag } from '@textkernel/oneui';

export default {
    title: 'Atoms/Tag',
    component: Tag,
    argTypes: {
        bgColor: {
            options: ['#3eff2b', '#ffa139', 'var(--color-background)'],
        },
        size: {
            options: ['small', 'normal', 'large'],
        },
    },
};

export const _Tag = (args) => (
    <div
        style={{
            padding: '5px',
        }}
    >
        <Tag {...args} />
    </div>
);
_Tag.args = {
    bgColor: '#3eff2b',
    isSelected: false,
    maxWidth: 'fit-content',
    size: 'normal',
    children: 'This is an extremely long long text!',
};

export const TagWithIcon = (args) => (
    <div
        style={{
            padding: '5px',
        }}
    >
        <Tag {...args} />
    </div>
);
TagWithIcon.args = {
    bgColor: '#3eff2b',
    isSelected: false,
    maxWidth: 'fit-content',
    size: 'normal',
    children: (
        <>
            <FiCheck />
            Some text
        </>
    ),
    contentClassName: 'test-class',
    contentStyle: {
        display: 'flex',
        alignItems: 'center',
        gap: 'var(--spacing-2x)',
    },
    onDelete: undefined,
};
