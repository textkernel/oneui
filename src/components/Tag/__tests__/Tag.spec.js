import React from 'react';
import toJson from 'enzyme-to-json';
import { Tag } from '../Tag';

describe('<Tag> component', () => {
    let wrapper;
    const onTagClick = jest.fn();
    const onDeleteClick = jest.fn();

    const deleteButtonSelector = 'Tag .Tag__deleteButton';

    const text = 'Tag text';
    const bgColor = '#ccc';
    const maxWidth = '30px';
    const textSize = 'large';

    afterEach(() => {
        jest.resetAllMocks();
    });

    it('should render correctly with minimum amount of props', () => {
        wrapper = mount(<Tag>{text}</Tag>);

        expect(wrapper.find('p.Tag__text').text()).toEqual(text);

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should render correctly with full list of props', () => {
        wrapper = mount(
            <Tag
                isSelected
                bgColor={bgColor}
                maxWidth={maxWidth}
                size={textSize}
                onClick={onTagClick}
                onDelete={onDeleteClick}
                contentClassName="my-class"
                contentStyle={{ color: 'red' }}
            >
                {text}
            </Tag>
        );

        expect(wrapper.find('.Tag').prop('className')).toContain('isSelected');
        expect(wrapper.find('.Tag').prop('style').backgroundColor).toEqual(bgColor);
        expect(wrapper.find('.Tag').prop('style').maxWidth).toEqual(maxWidth);
        expect(wrapper.find('Text').prop('size')).toEqual(textSize);
        expect(wrapper.find('Tag .Tag__deleteButton MdClose').exists()).toBeTruthy();
        expect(wrapper.find('.my-class')).toHaveLength(2);

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should invoke callback when onClick event is called', () => {
        wrapper = mount(<Tag onClick={onTagClick}>{text}</Tag>);

        wrapper.find('Tag').simulate('click');
        expect(onTagClick).toHaveBeenCalledTimes(1);
    });

    it('should invoke onClick when user interacts with tag by pressing Enter', () => {
        wrapper = mount(<Tag onClick={onTagClick}>{text}</Tag>);

        wrapper.find('Tag').simulate('focus');
        wrapper.find('Tag').simulate('keypress', { key: 'Enter' });

        expect(onTagClick).toHaveBeenCalledTimes(1);
    });

    it('should invoke callback when onDelete event is called', () => {
        wrapper = mount(
            <Tag onClick={onTagClick} onDelete={onDeleteClick}>
                {text}
            </Tag>
        );

        wrapper.find(deleteButtonSelector).simulate('click');
        expect(onTagClick).toHaveBeenCalledTimes(0);
        expect(onDeleteClick).toHaveBeenCalledTimes(1);
    });

    it('should invoke onDelete when user interacts with delete button by pressing Enter', () => {
        wrapper = mount(
            <Tag onClick={onTagClick} onDelete={onDeleteClick}>
                {text}
            </Tag>
        );

        wrapper.find(deleteButtonSelector).simulate('focus');
        wrapper.find(deleteButtonSelector).simulate('keydown', { key: 'Enter' });

        expect(onTagClick).toHaveBeenCalledTimes(0);
        expect(onDeleteClick).toHaveBeenCalledTimes(1);
    });
});
