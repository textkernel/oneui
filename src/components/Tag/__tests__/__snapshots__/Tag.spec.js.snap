// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tag> component should render correctly with full list of props 1`] = `
<Tag
  bgColor="#ccc"
  isSelected={true}
  maxWidth="30px"
  onClick={[MockFunction]}
  onDelete={[MockFunction]}
  size="large"
>
  <div
    className="Tag Tag--isSelected Tag--clickable"
    onClick={[MockFunction]}
    onKeyPress={[Function]}
    role="button"
    style={
      Object {
        "backgroundColor": "#ccc",
        "maxWidth": "30px",
      }
    }
    tabIndex={0}
  >
    <Text
      className="Tag__text"
      context="default"
      inline={false}
      size="large"
      title="Tag text"
    >
      <p
        className="Text--size_large Tag__text"
        title="Tag text"
      >
        Tag text
      </p>
    </Text>
    <button
      className="Tag__deleteButton"
      onClick={[Function]}
      onKeyPress={[Function]}
      type="button"
    >
      <MdClose
        size="15px"
      >
        <IconBase
          attr={
            Object {
              "viewBox": "0 0 24 24",
            }
          }
          size="15px"
        >
          <svg
            fill="currentColor"
            height="15px"
            size="15px"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width="15px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              key="0"
            />
          </svg>
        </IconBase>
      </MdClose>
    </button>
  </div>
</Tag>
`;

exports[`<Tag> component should render correctly with minimum amount of props 1`] = `
<Tag
  bgColor="var(--color-background)"
  isSelected={false}
  maxWidth="fit-content"
  size="normal"
>
  <div
    className="Tag"
    style={
      Object {
        "backgroundColor": "var(--color-background)",
        "maxWidth": "fit-content",
      }
    }
  >
    <Text
      className="Tag__text"
      context="default"
      inline={false}
      size="normal"
      title="Tag text"
    >
      <p
        className="Tag__text"
        title="Tag text"
      >
        Tag text
      </p>
    </Text>
  </div>
</Tag>
`;
